tutel/custom/custom_kernel.cpp

```diff
@@ -444,6 +460,54 @@ static void nccl_all_to_all_gather_async(
   g_cuda_events[0].record(get_nccl_stream());
 }
 
+static void all_to_all(torch::Tensor &output, torch::Tensor &input, const char *algo) {
+  CHECK_CUDA(output);
+  CHECK_CUDA(input);
+  auto recvbuff = (void*)output.data_ptr();
+  auto sendbuff = (void*)input.data_ptr();
+  cudaStream_t stream = get_nccl_stream().stream();
+
+  size_t length = input.nbytes();
+  CHECK_EQ(0, length % g_world_size);
+  size_t slice_size = length / g_world_size;
+
+  int nranks = g_world_size, ngpus = g_local_size;
+  CHECK_EQ(0, nranks % ngpus);
+  int nnodes = nranks / ngpus;
+  if (ngpus == 1 || nnodes == 1) goto linear;
+
+  if (algo && !strcmp(algo, "2D")) {
```

Comment:
We should add proper tensor buffer allocation hooks for send and recv buffer before this, else PyTorch tensor allocator might reuse these buffers before A2A is completed. Please reference all_to_all_gather_async and all_to_all_scatter_async.

Reply:
not true if not using multiple streams? the input/output are used in same stream currently